### DISCUSSION TITLE
Discussion: How recent are your reads?

### DISCUSSION THREAD

**Discussion by akoebbe** [tronikos/opower#18] [General]
One thing I was hoping for in this integration was to get near real-time (maybe 15-30min ago) usage data to see how specific household behaviors affect energy usage. At least for Evergy, the most recent read I can get is from the end of the previous day. I'm curious if this is an overall Opower limitation or if Evergy is sending Opower the day's meter reads only at the end of the day. How recent are the reads for everyone else?

**Comment by tronikos** [comment]
Similar for my utility PG&E. Typically at the end of the day I see previous day's usage/cost.

**Comment by akoebbe** [comment]
Interestingly the [Opower documentation](https://docs.oracle.com/en/industries/energy-water/opower-platform/data-transfer/interval_AMI_spec/DataIntegrationProcess_AMI_.htm) says daily uploads are a bare minimum.

> Interval data should be transferred as frequently as possible, but no less than on a daily basis to ensure timely display of data for online applications.

My hunch is the process for building the incremental upload files by the utilities is probably a resource intensive one, and doing it more frequently would result additional costs (assuming the jobs are cloud based). But this based solely on my assumptions from working in the server-side software industry.

I'm not sure how Oracle bills utilities for the Opower platform. I can't imagine it would include API call or data ingress frequency. But it is Oracle.

**Reply by tronikos** [reply]
Atlantic City Electric, see https://github.com/home-assistant/core/issues/97814, and First Energy, see https://github.com/tronikos/opower/discussions/19, only provide monthly data. ConEd, see https://github.com/tronikos/opower/issues/24 provides near-realtime (~4hr delay)! Hopefully other utilities will start providing near-realtime data.

**Reply by akoebbe** [reply]
Oh dang. I didn't realize how much this integration was taking off. You've probably been pretty busy. I appreciate your hard work.

I did try the real time URL posted in #24 (with the relevant bits changed to my provider) and only got the following message:
```
{"error":{"details":"No data returned from API. (24814820-b153-4b8d-8e97-64f4f79e5451)"}}
```

I'll repost it over there for posterity.

**Comment by tronikos** [comment]
I'm curious, you said you have quarter hour granularity. My electricity account on PG&E is supposed to have quarter hour too but I cannot get the data via their API. I can download using the green button but the data isn't really at quarter hour granularity. What it does for me is it divides the hourly data by 4 and provides the exact same value for each quarter of the hour. What about you?

**Reply by akoebbe** [reply]
When I downloaded the CSV from the "Green Button", this is what I get:

URL: https://kcpk.opower.com/ei/edge/apis/DataBrowser-v1/cws/utilities/kcpk/customers/[UUID]/usage_export/download?format=csv&startDate=2023-07-07&endDate=2023-08-07

```csv
...
TYPE,DATE,START TIME,END TIME,USAGE,UNITS,COST,NOTES
Electric usage,2023-07-06,00:00,00:14,0.96,kWh,$0.13,
Electric usage,2023-07-06,00:15,00:29,0.87,kWh,$0.11,
Electric usage,2023-07-06,00:30,00:44,0.99,kWh,$0.13,
Electric usage,2023-07-06,00:45,00:59,0.55,kWh,$0.07,
Electric usage,2023-07-06,01:00,01:14,0.56,kWh,$0.07,
Electric usage,2023-07-06,01:15,01:29,0.76,kWh,$0.10,
Electric usage,2023-07-06,01:30,01:44,0.40,kWh,$0.05,
Electric usage,2023-07-06,01:45,01:59,0.76,kWh,$0.10,
Electric usage,2023-07-06,02:00,02:14,0.25,kWh,$0.03,
...
```

Not sure if this is at all helpful but this is the response I get from `/ei/edge/apis/multi-account-v1/cws/kcpk/customers/current`

```json
{
    "id": "[5-digit int]",
    "uuid": "[UUID-1]",
    "legacyOpowerId": "[##-#-#aa#]",
    "accountNumber": "[My Billing Acct Num]",
    "accountName": "[empty]",
    "address": {
        "uuid": "[UUID-2]",
        "streetNumber": "[house number]",
        "streetName": "[street, all caps]",
        "subpremise": null,
        "postalCode": "[zip]",
        "city": "[city, all caps]",
        "country": "US",
        "state": "KS"
    },
    "type": "RESIDENTIAL",
    "utilityAccounts": [
        {
            "id": [6-digit-int-1],
            "uuid": "[UUID-3]",
            "utilityAccountId": "[10-digit int, Evergy meter id?]",
            "utilityAccountId2": "[10-digit int, Evergy meter id?]",
            "servicePointId": [6-digit-int-2],
            "meterType": "ELEC",
            "preferredUtilityAccountId": "[10-digit int, Evergy meter id?]",
            "readResolution": "QUARTER_HOUR"
        }
    ]
}
```

_Interesting note on the UUIDs_: Only the first segment of the UUIDs are different, the remaining 4 segments are the same for me. ex:
- UUID-1: `aaaaaaaa-1111-2222-3333-444444444444`
- UUID-2: `bbbbbbbb-1111-2222-3333-444444444444`
- UUID-3: `cccccccc-1111-2222-3333-444444444444`

**Reply by tronikos** [reply]
Interesting. For me it doesn't work well, see https://github.com/tronikos/opower/pull/29#issuecomment-1668930656
The information I get from customers/current is similar to yours.
